File: src/Simbuino.Emulator/Buttons.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Simbuino.Emulator
{
	// helper class for accessing the gamebuino buttons
	public class Buttons
	{
		PortRegister UP_PORT = AtmelContext.B; // D9
		const int UP_BIT = 1;
		PortRegister RIGHT_PORT = AtmelContext.D; // D7
		const int RIGHT_BIT = 7;
		PortRegister DOWN_PORT = AtmelContext.D; // D6
		const int DOWN_BIT = 6;
		PortRegister LEFT_PORT = AtmelContext.B; // D8
		const int LEFT_BIT = 0;
		PortRegister A_PORT = AtmelContext.D; // D4
		const int A_BIT = 4;
		PortRegister B_PORT = AtmelContext.D; // D2
		const int B_BIT = 2;
		PortRegister C_PORT = AtmelContext.C; // A3
		const int C_BIT = 3;

		public Buttons()
		{
			Reset();
		}

		public void Reset()
		{
			this.Up = false;
			this.Down = false;
			this.Left = false;
			this.Right = false;
			this.A = false;
			this.B = false;
			this.C = false;
		}

		public bool Up
		{
			get { return this.UP_PORT.ReadRegister[UP_BIT] == 0; }
			set { this.UP_PORT.ReadRegister[UP_BIT] = value ? 0 : 1; }
		}

		public bool Down
		{
			get { return this.DOWN_PORT.ReadRegister[DOWN_BIT] == 0; }
			set { this.DOWN_PORT.ReadRegister[DOWN_BIT] = value ? 0 : 1; }
		}

		public bool Left
		{
			get { return this.LEFT_PORT.ReadRegister[LEFT_BIT] == 0; }
			set { this.LEFT_PORT.ReadRegister[LEFT_BIT] = value ? 0 : 1; }
		}

		public bool Right
		{
			get { return this.RIGHT_PORT.ReadRegister[RIGHT_BIT] == 0; }
			set { this.RIGHT_PORT.ReadRegister[RIGHT_BIT] = value ? 0 : 1; }
		}

		public bool A
		{
			get { return this.A_PORT.ReadRegister[A_BIT] == 0; }
			set { this.A_PORT.ReadRegister[A_BIT] = value ? 0 : 1; }
		}

		public bool B
		{
			get { return this.B_PORT.ReadRegister[B_BIT] == 0; }
			set { this.B_PORT.ReadRegister[B_BIT] = value ? 0 : 1; }
		}

		public bool C
		{
			get { return this.C_PORT.ReadRegister[C_BIT] == 0; }
			set { this.C_PORT.ReadRegister[C_BIT] = value ? 0 : 1; }
		}

	}
}

File: src/Simbuino4Web/Simbuino4Web/Scripts/Views/Buttons.js
$(function () {

	// helper class for accessing the gamebuino buttons
	Buttons =
	{
		Init: function()
		{
			this.UP_PORT = AtmelContext.B; // D9
			this.UP_BIT = 1;
			this.RIGHT_PORT = AtmelContext.D; // D7
			this.RIGHT_BIT = 7;
			this.DOWN_PORT = AtmelContext.D; // D6
			this.DOWN_BIT = 6;
			this.LEFT_PORT = AtmelContext.B; // D8
			this.LEFT_BIT = 0;
			this.A_PORT = AtmelContext.D; // D4
			this.A_BIT = 4;
			this.B_PORT = AtmelContext.D; // D2
			this.B_BIT = 2;
			this.C_PORT = AtmelContext.C; // A3
			this.C_BIT = 3;

			this.Reset();
		},

		Reset: function ()
		{
			this.Up().set(false);
			this.Down().set(false);
			this.Left().set(false);
			this.Right().set(false);
			this.A().set(false);
			this.B().set(false);
			this.C().set(false);
		},

		Up: function()
		{
			var self = this;
			var result =
			{
				get: function () { return self.UP_PORT.ReadRegister.get().get_bit(self.UP_BIT) == 0; },
				set: function (value) { self.UP_PORT.ReadRegister.get().set_bit(self.UP_BIT, value ? 0 : 1); }
			};
			return result;
		},

		Down: function()
		{
			var self = this;
			var result =
			{
				get: function () { return self.DOWN_PORT.ReadRegister.get().get_bit(self.DOWN_BIT) == 0; },
				set: function (value) { self.DOWN_PORT.ReadRegister.get().set_bit(self.DOWN_BIT, value ? 0 : 1); }
			};
			return result;
		},

		Left: function()
		{
			var self = this;
			var result =
			{
				get: function () { return self.LEFT_PORT.ReadRegister.get().get_bit(self.LEFT_BIT) == 0; },
				set: function (value) { self.LEFT_PORT.ReadRegister.get().set_bit(self.LEFT_BIT, value ? 0 : 1); }
			};
			return result;
		},

		Right: function()
		{
			var self = this;
			var result =
			{
				get: function () { return self.RIGHT_PORT.ReadRegister.get().get_bit(self.RIGHT_BIT) == 0; },
				set: function (value) { self.RIGHT_PORT.ReadRegister.get().set_bit(self.RIGHT_BIT, value ? 0 : 1); }
			};
			return result;
		},

		A: function()
		{
			var self = this;
			var result =
			{
				get: function () { return self.A_PORT.ReadRegister.get().get_bit(self.A_BIT) == 0; },
				set: function (value) { self.A_PORT.ReadRegister.get().set_bit(self.A_BIT, value ? 0 : 1); }
			};
			return result;
		},

		B: function()
		{
			var self = this;
			var result =
			{
				get: function () { return self.B_PORT.ReadRegister.get().get_bit(self.B_BIT) == 0; },
				set: function (value) { self.B_PORT.ReadRegister.get().set_bit(self.B_BIT, value ? 0 : 1); }
			};
			return result;
		},

		C: function()
		{
			var self = this;
			var result =
			{
				get: function () { return self.C_PORT.ReadRegister.get().get_bit(self.C_BIT) == 0; },
				set: function (value) { self.C_PORT.ReadRegister.get().set_bit(self.C_BIT, value ? 0 : 1); }
			};
			return result;
		}
	};


});
